<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Project - Python Course</title>
    <link rel="stylesheet" href="css/styles.css">
</head>
<body>

<div class="navbar">
    <a href="index.html">Home</a>
    <a href="precourse.html">Precourse</a>
    <a href="lectures.html">Lectures</a>
    <a href="exercises.html">Exercises</a>
    <a href="quiz.html">Quiz</a>
    <a href="project.html">Project</a>
</div>


<div class="welcome-message">
<!-- project content goes here -->
    <h1>Introduction</h1>

    <p>After the lectures and exercises, you will have the opportunity to work on a course project. The project is designed to allow you to think more freely using the knowledge you have acquired, along with the tools we have covered during the lectures. It is a simple example of a problem you might encounter during your own research and is intended to prepare you to use Python to solve such problems.</p>

    <p>Please note that the project is not mandatory, and completing the course does not depend on its completion. However, we highly recommend that you attempt to solve the tasks and encourage you to ask questions if something is unclear. Since some of you might not have the time to complete the project during the course hours, we will be expecting questions regarding the project during the week following the course completion. </p>


    <h2>Background</h2>
    <p>For many diseases with known causative mutations, screening methods have been developed to detect whether people have a high risk of becoming sick, even before the onset of the actual disease. Over the last few years, the cost of full genome sequencing has gone down so that, in some cases, it might be cheaper to collect the complete genome sequence of patients with a high risk of carrying variants associated with the disease, rather than using targeted screening procedures.</p>

    <p>Cystic fibrosis is a complex disease, where patients often manifest the following symptoms: problems with lung functions, diabetes and infertility. From a genetic point of view, there are several mutations associated with this disease. The gene CFTR (short for Cystic Fibrosis Transmembrane Conductance Regulator) encodes an ion channel protein acting in epithelial cells and is encoded on chromosome 7 of the human genome. CFTR carries several non-synonymous genetic variants, some of which leading to premature stop codons that are known to cause the disease.</p>

    <p>In the annotation file (the GTF file), the CFTR gene has the id <b>ENSG00000001626</b> on the chromosome <b>7</b>, and this entry can be found in the <b>attribute</b> field in the GTF file</p>

    <h2>Goal</h2>
    <p>In this assignment, you have access to the human reference genome (chromosome 7) as well as the full genome annotation. In addition, you have genome sequence data of chromosome 7 from five individuals from a family at risk of carrying mutations related to the disease.</p>
    <p>Your task is to write a Python program that:</p>
    <ol>
        <li>extracts a disease-causing transcript from the CFTR gene,</li>
        <li>translates the gene sequence to its corresponding amino-acid sequence, and</li>
        <li>based on the reference amino-acid sequence determine whether any of the five given individuals is affected.</li>
    </ol>

    <h2>Required Material</h2>
    <p>The main task is divided into several steps. The first step is to download the <a href="assignment/data.zip">data.zip</a> which includes
    <table class="filetable">
        <tr>
            <th>Filename</th>
            <th>Description</th>
        </tr>
        <tr>
            <td>Homo_sapiens.GRCh38.93.gtf</td>
            <td>Human reference annotation file (GTF format)</td>
        </tr>
        <tr>
            <td>Homo_sapiens.GRCh38.dna_sm.chromosome.7.fa</td>
            <td>Human reference DNA sequence for chromosome 7 (fasta format)</td>
        </tr>
        <tr>
            <td>Patient1.fa</td>
            <td>DNA sequence data for Patient 1 (fasta format)</td>
        </tr>
        <tr>
            <td>Patient2.fa</td>
            <td>DNA sequence data for Patient 2 (fasta format)</td>
        </tr>
        <tr>
            <td>Patient3.fa</td>
            <td>DNA sequence data for Patient 3 (fasta format)</td>
        </tr>
        <tr>
            <td>Patient4.fa</td>
            <td>DNA sequence data for Patient 4 (fasta format)</td>
        </tr>
        <tr>
            <td>Patient5.fa</td>
            <td>DNA sequence data for Patient 5 (fasta format)</td>
        </tr>
        <tr>
            <td>gtf-format.md</td>
            <td>Markdown file explaining the GTF format</td>
        </tr>
    </table>



    <p>If you are not familiar with the file formats, read up online on how the files are structured. For example, <a href="assignment/data/gtf-format.md" target="_blank">here</a> you can find a short description of the different (tab-delimited) fields of a GTF file and <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=BlastHelp" target="_blank">here</a> a general description about the fasta files.</p>
    <p>Some of the tasks involve outputting long sequences. To make sure they are correct, use the <code>utils.check_answers</code> package (download from <a href="assignment/utils.zip" target="_blank">utils.zip</a>). You can import it that way:</p>
    <pre><code>from utils import check_answers</code></pre>
    <p>More detailed instructions are given with each task that uses the package.</p>

    <h2>Warm-up</h2>
    <p>We recommend creating a folder in your file system, called <b>project</b>. In this folder you can add all the required files, such as the reference and annotation files, as well as the downloaded folder from the previous step.</p>
    <ol>
        <li>Unzip the reference DNA and annotation files (fasta and GTF respectively), so that you can check their contents. That should be possible using 7zip in Windows and double clicking the files in MacOS and Linux.</li>
        <p>The instructions below works for Mac or Linux users</p> 
        <li>Now that the files are decompressed, you can check their content using the <code>head</code> command on your terminal. You can search online how to use the <code>head</code> command or type <code>head --help</code> to see the usage instructions. You can check the first two lines of each file using the <code>head</code> command.</li>
    </ol>


    <div class="question">
      <div class="toggle-btn" onclick="toggleContent(this, 'answer-content')">
          <p><strong>Answer:</strong></p>
      </div>
      <div class="answer-content">
          <p>For Linux or Mac users</p>
          <p>Run <code>head -n 2</code> to see the first two lines</p>
          <p>For Windows user, please use Python to loops through the file by line and print only the first two lines, i.e. <b>break</b> the loop when the line counter is more than 2.</p>
      </div>
    </div>

    <ol start="3">
        <li>What is the length of chromosome 7 on the reference sequence?</li>
    </ol>

    <div class="question">
      <div class="toggle-btn" onclick="toggleContent(this, 'tip-content')">
        <p><strong>Tip:</strong></p>
      </div>
      <div class="tip-content">
        <p>Open the reference fasta file and read it line by line. In a loop, ignore the first line and get the length of each following line. Donâ€™t forget to remove the trailing newline character from each line. Sum up all the lengths you found.</p>
        <p>You can also use the Python script <b>cal_seqlen.py</b> you created from the 4th question in <a href="https://python-bioinfo.bioshu.se/publish/Day_3_Exercise_2.html" target="_blank">Day 3, Exercise 2</a></p>
      </div>
    </div>

    <div class="question">
      <div class="toggle-btn" onclick="toggleContent(this, 'answer-content')">
        <p><strong>Answer:</strong></p>
      </div>
      <div class="answer-content">
        <p>Chromosome 7 has 159,345,973 base pairs.</p>
      </div>
    </div>

    <ol start="4">
        <li>How many genes are annotated in the GTF file?</li>
    </ol>
    <div class="question">
      <div class="toggle-btn" onclick="toggleContent(this, 'tip-content')">
        <p><strong>Tip:</strong></p>
      </div>
      <div class="tip-content">
        <p>You need to understand the structure of a GTF (gene transfer format) file for this project. Take your time and read up on the file format if you are not sure how to solve this task. To get the number of genes, open the GTF file and read it line by line. In a loop, count all features of type <b>gene</b>.</p>
        <p><b>feature</b> is the second column of a line, which is separated by tabs (<b>'\t'</b>)</p>
        <p>To divide a string with items separated by tabs, using
          <pre><code>entries = line.split('\t')</code></pre>
        </p>
      </div>
    </div>

    <div class="question">
      <div class="toggle-btn" onclick="toggleContent(this, 'answer-content')">
        <p><strong>Answer:</strong></p>
      </div>
      <div class="answer-content">
        <p>There are 58,395 genes annotated in the GTF file.</p>
      </div>
    </div>

    <h2>Architect a Method</h2>
    <p>All following tasks are related to the CFTR gene.</p>
    <p>In the annotation file (the GTF file), the CFTR gene has the id <b>ENSG00000001626 </b> on the chromosome 7.</p>
    <ol>
        <li>How many transcripts can the CFTR gene generate?</li>
    </ol>

    <div class="question">
      <div class="toggle-btn" onclick="toggleContent(this, 'tip-content')">
        <p><strong>Tip:</strong></p>
      </div>
      <div class="tip-content">
        <p>Again, think about the structure of the GTF file. Open the GTF file. In a loop, count all transcript features for the gene.</p>
        <p>Loop over all lines, first check if the <b>feature</b> (3rd column of the line) is equal to <b>"transcript"</b>. Then check if the <b>gene_id</b> of the <b>attribute</b> field (the 9th column) is equal to <b>ENSG00000001626</b></p>
        <p>The attribute field is separated by <b>';'</b>, and <b>gene_id</b> is located in the first element of the attribute field. <b>gene_id</b> can be extracted from the attribute field by 
        <pre><code>
li_attr = attribute.split(";")
gene_id = li_attr[0].strip().split()[1].strip('"')
        </code></pre></p>
        <p>You can also use the following helper function to parse the attribute filed into a dictionary and that solution is more robust.
        <pre><code>
def parse_attribute(attribute):
    """
    Parse the attribute field from the GTF file
    Input: string of the attribute field 
    Output: a dictionary
    """
    li_attr = [x.strip() for x in attribute.split(';') if x.strip()]
    d_attr = {}
    for attr in li_attr:
        key_and_value = attr.split(' ', maxsplit=1)
        if (len(key_and_value) < 2):
            print(key_and_value)
        key = key_and_value[0]
        value = key_and_value[1].strip().strip('"')
        d_attr[key] = value
    return d_attr

# example usage 
attribute = 'gene_id "ENSG00000001626"; gene_version "14"; transcript_id "ENST00000546407"; transcript_version "1"; gene_name "CFTR"; gene_source "ensembl_havana"; gene_biotype "protein_coding"; transcript_name "CFTR-207"; transcript_source "havana"; transcript_biotype "processed_transcript"; transcript_support_level "1";'
d_attr = parse_attribute(attribute)
print(d_attr)
gene_id = d_attr['gene_id']
transcript_id = d_attr['transcript_id']
         </code></pre>

        </p>
      </div>
    </div>

    <div class="question">
      <div class="toggle-btn" onclick="toggleContent(this, 'answer-content')">
        <p><strong>Answer:</strong></p>
      </div>
      <div class="answer-content">
        <p>The CFTR gene can produce 11 different transcripts.</p>
      </div>
    </div>

    <ol start="2">
        <li>Which of these transcripts is the longest transcript in nucleotides?
          Print the transcript id and its length in nucleotides.
        </li>
    </ol>

    <div class="question">
      <div class="toggle-btn" onclick="toggleContent(this, 'tip-content')">
        <p><strong>Tip:</strong></p>
      </div>
      <div class="tip-content">
        <p>Open the GTF file. Fetch the start (4th column) and stop (5th column) positions for each transcript of the CFTR gene to calculate its length. Keep in mind that sequence numbering starts at 1 in the GTF file format. Therefore, you need to subtract the start position from the stop position and add 1 to get the length of the transcript.
          <pre><code>
length = stop_position - start_position + 1
          </code></pre>
        </p>
        In order to get the transcript with the longest length, you can use
        two a tuple to store the transcript id and its length. Then, in each iteration, compare the length of the current transcript with the length of the longest transcript. If the current transcript is longer, update the tuple with the current transcript id and its length. 
        <p>
        </p>

        </p>
      </div>
    </div>


    <div class="question">
      <div class="toggle-btn" onclick="toggleContent(this, 'answer-content')">
        <p><strong>Answer:</strong></p>
      </div>
      <div class="answer-content">
        <p>The transcript with the id ENST00000003084 is the longest of 11 transcripts and spans 188,703 bases.</p>
      </div>
    </div>

    <ol start="3">
        <li>Fetch the DNA sequence for that transcript.</li>
    </ol>

    <div class="question">
      <div class="toggle-btn" onclick="toggleContent(this, 'tip-content')">
        <p><strong>Tip:</strong></p>
      </div>
      <div class="tip-content">
        <p>Open the reference fasta file. In a loop, ignore the first line and append each line to a list, removing the trailing newline character. Outside the loop, use the join function to concatenate the lines from the list to get the reference sequence. Avoid concatenation inside the loop, as it is slow and wastes memory. Extract the start and stop positions of the longest transcript to fetch its DNA sequence from the reference sequence, but think about where the index starts from.save the result in a string called e.g. <b>ref_seq</b>.</p>
        <p>You can also use the function <b>readseq(seqfile)</b> from the 2nd question of <a href="https://python-bioinfo.bioshu.se/publish/Day_3_Exercise_2.html" target="_blank">Day 3, Exercise 2</a>, although it's not optimal</p>
        <p>Fetch the start and stop position for the longest transcript (<b>ENST00000003084</b>)
          Then use the following code to extract the DNA sequence of the transcript
          <pre><code>
start = 117 # replace this with the actual start position
stop = 188 # replace this with the actual stop position
transcript_seq = ref_seq[start-1:stop]
          </code></pre>
        </p>
      </div>
    </div>

    <div class="question">
      <div class="toggle-btn" onclick="toggleContent(this, 'answer-content')">
        <p><strong>Answer:</strong></p>
      </div>
      <div class="answer-content">
        <p>Write your results to a file and compare it to the correct result using <code>check_answers.ex3("resultsFile.txt")</code>. The entire sequence can also be found <a href="assignment/results/transcript.ncbi.fasta" target="_blank">here</a>.</p>
      </div>
    </div>

    <ol start="4">
        <li>Fetch the DNA sequences of all exons for that transcript, concatenate together to one sequence.</li>
    </ol>

    <div class="question">
      <div class="toggle-btn" onclick="toggleContent(this, 'tip-content')">
        <p><strong>Tip:</strong></p>
      </div>
      <div class="tip-content">
        <p>First, you need to save the start and stop positions of all exons of the longest transcript. Then you can use a similar loop to the one you used in task 3 to extract the DNA sequence of each exon. Finally, you need to concatenate the DNA sequences of all exons.</p>
        <p>An exon is recognized if the feature (3rd column of the line) is equal to <b>exon.</b></p>
        <p>Once you have retrieve start and stop positions for all exons, use the following code to extract the DNA sequence of the exons
          <pre><code>
exon_seq = ""
for start, stop in exon_positions:
    exon_seq += ref_seq[start-1:stop]
          </code></pre>
          Or you can also add the sequences of each exon to a list and then join them together.
        </p>
      </div>
    </div>


    <div class="question">
      <div class="toggle-btn" onclick="toggleContent(this, 'answer-content')">
        <p><strong>Answer:</strong></p>
      </div>
      <div class="answer-content">
        <p>Write your results to a file and compare it to the correct result using <code>check_answers.ex4("resultsFile.txt")</code>. The correct sequence can also be found <a href="assignment/results/mrna.ncbi.fasta" target="_blank">here</a>.</p>
      </div>
    </div>

    <ol start="5">
        <li>What are the start positions and sequences of the start_codon and stop_codon for that transcript?</li>
    </ol>


    <div class="question">
      <div class="toggle-btn" onclick="toggleContent(this, 'tip-content')">
        <p><strong>Tip:</strong></p>
      </div>
      <div class="tip-content">
        <p>Find the start_codon and stop_codon features of the longest transcript in the GTF file, including the start positions of the start- and stop-codon. Check that the start_codon has the sequence ATG, and that the stop_codon corresponds to a proper stop codon. Make your program print a warning message in case the transcript does not begin with a start-codon and end with a stop-codon.</p>
        <p>When <b>feature</b>(the 3rd column) is equal to <b>start_codon</b>, the start position (4th column) is the position of the start codon</p>
        <p>When <b>feature</b>(the 3rd column) is equal to <b>stop_codon</b>, the start position (4th column) is the position of the stop codon</p>
        <p>The start codon (3 letter nucleotides) can be retrieved by the following code
          <pre><code>
start_codon = ref_seq[pos_start_codon-1:pos_start_codon+2]
          </code></pre>
        <p>The stop codon (3 letter nucleotides) can be retrieved by the following code
          <pre><code>
stop_codon = ref_seq[pos_stop_codon-1:pos_stop_codon+2]
          </code></pre>
          <p>If start_codon is equal to <b>ATG</b>, its good, otherwise print the warning message</p>
          <p>If stop_codon is one of <b>[TAG, TAA, TGA]</b>, its good, otherwise print the warning message</p>
        </p>
      </div>
    </div>

    <div class="question">
      <div class="toggle-btn" onclick="toggleContent(this, 'answer-content')">
        <p><strong>Answer:</strong></p>
      </div>
      <div class="answer-content">
        <p>The start codon has the sequence ATG and starts at position 117,480,095.</p>
        <p>The stop codon has the sequence TAG and starts at position 117,667,106.</p>
      </div>
    </div>

    <ol start="6">
        <li>Translate the above sequence of all exons into amino acids, using an implementation of the translation table from the <code>utils.rna</code> package (included in the `utils` folder).</li>
    </ol>


    <div class="question">
      <div class="toggle-btn" onclick="toggleContent(this, 'tip-content')">
        <p><strong>Tip:</strong></p>
      </div>
      <div class="tip-content">
        <p>Translate the DNA sequence of the concatenated exons into amino acids from the start codon position of the transcript on.</p>
        <p>
          Suppose <b>pos_exons</b> contains the start and end positions of all exons of the transcript, <b>pos_start_codon</b> is the start position of the start codon of the transcript, and <b>mRNA</b> is the concatenated sequence of all exons. Then you can use the following code to translate the DNA sequence of the exons into amino acids
          <pre><code>
from utils.rna import translate_dna
pos_start = pos_start_codon - exons[0][0] # the start position of the first exon of the mRNA
aminoacids = translate_dna(mRNA[start:])
          </code></pre>
        </p>
      </div>
    </div>

    <div class="question">
      <div class="toggle-btn" onclick="toggleContent(this, 'answer-content')">
        <p><strong>Answer:</strong></p>
      </div>
      <div class="answer-content">
        <p>Write your results to a file and compare it to the correct result using <code>check_answers.ex6("resultsFile.txt")</code>. The correct sequence can also be found <a href="assignment/results/protein.ncbi.fasta" target="_blank">here</a>.</p>
      </div>
    </div>

    <h2>Find the Patients at Risk</h2>
    <p>We are reaching the goal for this assignment!</p>
    <p>A mutation in the transcript ENST00000003084 causes a premature stop codon to be introduced into the amino acid sequence. This creates a truncated protein, causing cystic fibrosis. Use the python code you have written to solve the tasks above and extend it to compare the reference genome sequence of chromosome 7 to the sequences of the following 5 patients in fasta format (Patient1.fa.gz, Patient2.fa.gz, Patient3.fa.gz, Patient4.fa.gz, and Patient5.fa.gz) to determine which of them are carrying a mutation in the CFTR gene that causes a truncated protein.</p>

    <h2>Extra Task</h2>
    <ol>
        <li>Use BioPython to parse the fasta file and to translate DNA nucleotides into amino acids.</li>
    </ol>
    <div class="question">
      <div class="toggle-btn" onclick="toggleContent(this, 'tip-content')">
        <p><strong>Tip:</strong></p>
      </div>
      <div class="answer-content">
        <p>Check the BioPython tutorial on how to parse a fasta file with BioPython. Read up on the built-in translation method and the BioPython translation tables.</p>
      </div>
    </div>

</div>


<!-- Script to handle toggling -->
<script>
function toggleContent(btn, className) {
    var content = btn.nextElementSibling;
    if (content.style.display === "none" || content.style.display === "") {
        content.style.display = "block";
        btn.classList.add('expanded');
    } else {
        content.style.display = "none";
        btn.classList.remove('expanded');
    }
}

function toggleSection(element) {
    var content = element.nextElementSibling;
    if (content.style.display === "none" || !content.style.display) {
        content.style.display = "block";
        element.querySelector('.arrow').style.transform = 'rotate(-135deg)';
    } else {
        content.style.display = "none";
        element.querySelector('.arrow').style.transform = 'rotate(45deg)';
    }
}
</script>

</body>
</html>
